### PR TITLE
Add: check permissions for session key owners in validation

### PR DESCRIPTION
**Pull request by sm-stack**

### CODE REVIEW

File: contracts/test/TestExpiryAccount.sol [hunk 1]
```diff
@@ -42,15 +42,54 @@ contract TestExpiryAccount is SimpleAccount {
...
+        
+        // we have "until" value for all valid owners. so zero means "invalid signature"
         bool sigFailed = _until == 0;
-        return _packValidationData(sigFailed, _until, _after);
+        if (sigFailed) {
+            return _packValidationData(sigFailed, _until, _after);
+        }
+        
+        // All external function call is made through execute(address dest, uint256 value, bytes calldata func) at SimpleAccount.sol 
+        if (getSelector(calldataCopy) != "0xdade6037") {
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Please also add executeBatch function here!

Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Please change calldataCopy to userOp.calldata

File: contracts/test/TestExpiryAccount.sol [hunk 2]
```diff
@@ -42,15 +42,54 @@ contract TestExpiryAccount is SimpleAccount {
...
+        (address dest, uint256 value, bytes memory func) = decode(userOp.calldata);
+
+        uint256 permissionLength = permissions.length;
+        for(uint i; i < permissionLength; i++) {
+            PermissionParam memory permission = permissions[i];
+            if (permission.whitelistDestination == dest) {
+                uint256 permissionMethodsLength = permission.whitelistMethods.length;
+                if (permissionMethodsLength > 0) {
+                    for(uint j; j < permissionMethodsLength; j++) {
+                        if (permission.whitelistMethods[j] == getSelector(userOp.calldata)) {
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: userOp.calldata -> func

File: contracts/test/TestExpiryAccount.sol [hunk 3]
```diff
@@ -42,15 +42,54 @@ contract TestExpiryAccount is SimpleAccount {
...
+        if (getSelector(calldataCopy) != "0xdade6037") {
+            sigFailed = true;
+            // _packValidationData defined at core/Helper.sol
+            return _packValidationData(sigFailed, _until, _after);
+        }
+
+        (address dest, uint256 value, bytes memory func) = decode(userOp.calldata);
+
+        uint256 permissionLength = permissions.length;
+        for(uint i; i < permissionLength; i++) {
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Should consider changed addTemporaryOwner function

**Comment by sm-stack** [comment]
Fixed all above checks
Resolved type collision at getSelector / addTemporaryOwner
Fixed invalid sigFailed flag at _validateSessionKey

File: contracts/test/TestExpiryAccount.sol [hunk 4]
```diff
@@ -39,18 +58,109 @@ contract TestExpiryAccount is SimpleAccount {
...
+            }
+        }
         return _packValidationData(sigFailed, _until, _after);
     }
+
+    function getSelector(bytes calldata _data) public pure returns (bytes4 selector) {
+        selector = bytes4(_data[0:4]);
+    }
+
+    function decodeSingle(bytes calldata _data) public pure returns (address dest, uint256 value, bytes memory func){
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Please change function name to _decodeSingle!

File: contracts/test/TestExpiryAccount.sol [hunk 5]
```diff
@@ -39,18 +58,109 @@ contract TestExpiryAccount is SimpleAccount {
...
+        (dest, value, func) = abi.decode(_data[4:], (address, uint256, bytes));
+    }
+
+    function decodeBatch(bytes calldata _data) public pure returns (address[] memory dest, bytes[] memory func){
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Please change function name to _decodeBatch!

File: contracts/test/TestExpiryAccount.sol [hunk 6]
```diff
@@ -39,18 +58,109 @@ contract TestExpiryAccount is SimpleAccount {
...
+            if (permissionStorage.whitelistMethodsMap[dest][this.getSelector(func)]) {
+                    sigFailed = false;
+                    return _packValidationData(sigFailed, _until, _after);
+            } 
+        }
+        return _packValidationData(sigFailed, _until, _after);
+        
+    }
+
+    function _validateSessionKeyBatch(bytes calldata userOpCallData, address signer, uint48 _until, uint48 _after)
```
Comment (Sangyeup) [inline contracts/test/TestExpiryAccount.sol]: Will be better if we can integrate _validateSessionKeySingle and _validateSessionKeyBatch to one function. Below is what I changed! You should also change _validateSignature and _decodeSingle function to adjust.

`
function _validateSessionKey(
    bytes calldata userOpCallData, address signer, uint48 _until, uint48 _after, bytes4 userOpSelector
)

    internal view returns (uint256 validationData) {

	if (userOpSelector == FUNCTION_EXECUTE) {
	    (address[] memory dest, , bytes[] memory func) = decodeSingle(userOpCallData); 	
	} else {
	    (address[] memory dest, bytes[] memory func) = decodeBatch(userOpCallData);
	}

        PermissionStorage storage permissionStorage = permissionMap[signer];

        bool sigFailed = true;

        uint256 length = dest.length;
        for (uint256 i = 0; i < length; i++) {
            if (permissionStorage.whitelistDestinationMap[dest[i]]) {
                if (permissionStorage.whitelistMethodsMap[dest[i]][this.getSelector(func[i])]) {
                    sigFailed = false;
                    break;
                    // Token Approval??
                }
            }
        }
        return _packValidationData(sigFailed, _until, _after);`

**Review by Sangyeup — CHANGES_REQUESTED** [review]

**Comment by Sangyeup** [comment]
@sm-stack LGTM!

**Pull request merged**